### PR TITLE
Linear probe in sets only when appropriate.

### LINKED ISSUE #6543: Operation "element in set" never returns
Bug: After a series of add and remove operations on a set in nopython mode the operation "in set" does not return and gets stuck at 100% CPU load.

Isolated code example:
```python
import numba as nb

@ nb.njit()
def problem():

    # init
    openSet = set()

    openSet.add((0, 0))
    openSet.add((1, 0))
    openSet.add((2, 0))
    openSet.add((3, 0))
    openSet.add((4, 0))
    openSet.add((5, 0))
    openSet.add((6, 0))
    openSet.add((7, 0))
    openSet.add((8, 0))

    openSet.remove((2, 0))
    openSet.add((2, 1))
    openSet.remove((2, 1))
    openSet.remove((4, 0))
    openSet.add((5, 1))
    openSet.add((4, 1))
    openSet.remove((7, 0))
    openSet.add((8, 1))
    openSet.add((7, 1))
    openSet.add((6, 1))
    openSet.remove((8, 0))
    openSet.remove((1, 0))
    openSet.remove((0, 0))
    openSet.remove((5, 0))
    openSet.remove((6, 0))
    openSet.remove((3, 0))
    openSet.remove((7, 1))
    openSet.add((8, 2))
    openSet.add((7, 2))
    openSet.add((6, 2))
    openSet.remove((8, 1))
    openSet.remove((6, 1))
    openSet.add((5, 2))
    openSet.remove((4, 1))
    openSet.add((4, 2))
    openSet.remove((5, 1))
    openSet.remove((6, 2))
    openSet.add((7, 3))
    openSet.add((6, 3))
    openSet.add((5, 3))
    openSet.remove((7, 2))
    openSet.add((8, 3))
    openSet.remove((8, 2))
    openSet.remove((4, 2))
    openSet.add((4, 3))
    openSet.add((3, 3))
    openSet.remove((5, 2))
    openSet.remove((7, 3))
    openSet.add((8, 4))
    openSet.add((7, 4))
    print('About to check (6,4) in openSet')
    return (6, 4) in openSet


if __name__ == "__main__":
    print(problem())
    print('DONE')
```

The code gets stuck with numba 0.52.0 with Python 3.8.5 on Ubuntu 20.04.1 LTS x86_64.

However, @esc on gitter.im also confirmed that the issue persists on OSX.

Isolated code example:
https://github.com/icezyclon/astart-pathfinding/blob/main/isolated.py
Original code example (from A* search):
https://github.com/icezyclon/astart-pathfinding/blob/main/problem.py

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #6543

### CODE REVIEW

File: numba/cpython/setobj.py [hunk 1]
```diff
@@ -254,7 +254,12 @@ def check_entry(i):
         # First linear probing.  When the number of collisions is small,
         # the lineary probing loop achieves better cache locality and
         # is also slightly cheaper computationally.
-        with cgutils.for_range(builder, ir.Constant(intp_t, LINEAR_PROBES)):
+        ll_lp = ir.Constant(intp_t, LINEAR_PROBES)
+        pred = builder.icmp_unsigned("<=",
+                                     builder.add(ll_lp, builder.load(index)),
+                                     mask)
+        probes = builder.select(pred, ll_lp, ir.Constant(intp_t, 0))
```
Comment (esc) [inline numba/cpython/setobj.py:261]: perhaps define `zero` above, where `one` and `five` are defined?

**Comment by benwilliamgraham** [comment]
Hi @stuartarchibald, do you mind if I finish up this pr with https://github.com/numba/numba/pull/7875?

**Comment by stuartarchibald** [comment]
I'm unconvinced that this totally fixes the problem. So am going to close this, the patch is still present if it's deemed useful.

**Comment by esc** [comment]
I think this was closed in error.

**Comment by stuartarchibald** [comment]
> I think this was closed in error.

@esc I deliberately closed it, xref discussion at: https://github.com/numba/numba/pull/7875#issuecomment-1056730734. Am happy to reopen if you think otherwise!

**Comment by esc** [comment]
@stuartarchibald IIRC this was part of the solution and fixed, but I see now that your original PR was "adopted", so it can indeed be left closed, sorry for the noise.

**Comment by stuartarchibald** [comment]
> @stuartarchibald IIRC this was part of the solution and fixed, but I see now that your original PR was "adopted", so it can indeed be left closed, sorry for the noise.

@esc No problem, from looking at the previous discussion it seemed like it was a part fix, but even with another part fix from else where there were still issues. The main concern I had (and still have) is that it's not clear what the actual problem is or why these things help. I'll leave this closed for now, hopefully a `typed.Set` implementation wouldn't have such issues!

**Pull request closed without merging**